helpers/MigrationHelper.ts

```diff
@@ -0,0 +1,246 @@
+        contractMeta.address,
+        this._signer
+      );
+    }
+
+    throw new Error("Invalid contract meta");
+  }
+
+  @log
+  async readEventLogData(
+    contract: Contract,
+    firstEventBlock: number,
+    eventsChunkSize: number,
+    eventName = "StateUpdated",
+    fileName = "events-data.json"
+  ): Promise<any[]> {
+    const filter = contract.filters[eventName](null, null, null, null);
+    const latestBlock = await ethers.provider.getBlock("latest");
+    console.log("startBlock", firstEventBlock, "latestBlock Number", latestBlock.number);
+
+    let logHistory: unknown[] = [];
+
+    for (let index = firstEventBlock; index <= latestBlock.number; index += eventsChunkSize) {
+      let pagedHistory;
+      try {
+        const toBlock =
+          index + eventsChunkSize > latestBlock.number
+            ? latestBlock.number
+            : index + eventsChunkSize - 1;
+        pagedHistory = await contract.queryFilter(filter, index, toBlock);
+      } catch (error) {
+        console.error(error);
+      }
+      console.log(
+        `state transition history length: ${pagedHistory.length}, current block number: ${index}, latest block number: ${latestBlock.number}`
+      );
+      logHistory = [...logHistory, ...pagedHistory];
+    }
+    console.log(`Total events count: ${logHistory.length}`);
+
+    // save data to file
+    if (fileName) {
+      this.writeFile(fileName, logHistory);
+    }
+
+    return logHistory;
+  }
+
+  @log
+  async migrateData(
+    data: EventLogEntry[],
+    populateDataFn: (...args) => Promise<any>,
+    fileName = "migration-result.json"
+  ): Promise<MigrationResult> {
+    const result: MigrationResult = {
+      migratedData: [],
+      receipt: null,
+      error: null,
+      index: 0,
+    };
+    for (let idx = 0; idx < data.length; idx++) {
+      result.index = idx;
+      try {
+        const args = data[idx].args;
+        const receipts = await populateDataFn(args);
+        for (const receipt of receipts) {
+          result.migratedData.push({
+            args,
+            tx: receipt.transactionHash,
+          });
+          if (receipt.status !== 1) {
+            result.receipt = receipt;
+            result.error = "receipt status failed";
+            break;
+          }
+        }
+      } catch (error) {
+        console.error(error);
+
+        result.error =
+          typeof error === "string"
+            ? error
+            : JSON.stringify(error, Object.getOwnPropertyNames(error));
+        console.log("migration error", result.error, result.receipt);
+
+        this.writeFile(fileName, result);
+        throw error;
+      }
+    }
+
+    console.log("migration completed successfully");
+    this.writeFile(fileName, result);
+
+    return result;
+  }
+
+  protected writeFile(fileName: string, data: unknown): void {
+    fs.writeFileSync(fileName, toJson(data));
+  }
+}
+
+export class StateTestContractMigrationSteps extends ContractMigrationSteps {
```

Comment:
May be it worth separate the concrete ContractMigrationSteps implementation from its abstract class by moving it to other file?